### PR TITLE
feat(path): single file exports

### LINKED ISSUE #687: path: Make os an option argument
The export structure of the `path` module looks like this:
```ts
export const posix = {
  basename,
  dirname,
  extname,
  ...
};

export const win32 = {
  basename,
  dirname,
  extname,
  ...
};

export {
  basename,
  dirname, 
  extname,
  ...,
  globToRegExp,
  normalizeGlob,
  SEP,
  SEP_PATTERN,
  ...
};
```

This seems archaic and overly bound to where we vendored the implementation. It forces us to put group every function by OS in just `path/posix.ts` and `path/win32.ts`, rather than having modules like `path/join.ts`, `path/file_url.ts`, etc. It's hard to maintain and the test modules don't match it.

We should instead export one set of functions which each have an options interface containing the OS. They all have few enough positional arguments. This is modelled already by `path/glob.ts`, which gets to have its own module, containing:
```ts
export interface GlobOptions {
  ...,
  /** Operating system. Defaults to the native OS. */
  os?: typeof Deno.build.os;
}
```
These functions can branch by OS individually.

The current structure allows importing one namespace of functions for another OS, but this is niche enough that users should make such an abstraction themselves if they need it (in reality they can just import it from the `node/` polyfill).

Ref https://github.com/denoland/deno/discussions/8405.

### PR DESCRIPTION
**Pull request by lino-levan**
Related #687.

Future PRs:
- BREAKING: Add "os" argument
- BREAKING: Rename some methods to make them more intuitive (by @nayeemrmn on discord)
  - basename() -> fileName()
  - dirname() -> parent()
  - extname() -> extension()

### CODE REVIEW

File: path/dirname.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  }
+  return stripTrailingSeparators(path.slice(0, end), isPosixPathSeparator);
+}
+
+/**
+ * Return the directory path of a `path`.
+ * @param path - path to extract the directory from.
+ */
+export function dirname(path: string): string {
+  assertPath(path);
```
Comment (lino-levan) [inline path/dirname.ts:132]: comment: found a bug in our current posix implementation of `dirname` of it not asserting the path at the start (the win32 version did that)

**Comment by kt3k** [comment]
Move to single export file structure looks good to me, and it's probably less controversial to other points if it doesn't have breaking changes. Can we only do the move to single-file export without API design changes in this PR?

> Add "os" argument
> Make some breaking changes to APIs to allow for more options in the future

These points seem needing more discussion, also these 2 changes don't seem necessarily in the same PR. Maybe try them later PRs?

**Comment by lino-levan** [comment]
I guess you're right, but I just wanted to bundle breaking changes together. I'm not sure what the best way to move forward with single-file exports is. Should we expose `posixX()` and `windowsX()` to the user? Does that naming make any sense?

**Comment by kt3k** [comment]
I think we can also update `mod.ts` to use new implementations and that should make the rendering of [this page](https://deno.land/std@0.196.0/path/mod.ts) better.

> Should we expose posixX() and windowsX() to the user?

I don't think so. The names sound strange to me.

I don't think we should deprecate `win32.ts` and `posix.ts` in this PR. I think that should be discussed together with `Add "os" argument` issue

**Comment by kt3k** [comment]
Also can you add the license headers `// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.` to the new files?

**Comment by lino-levan** [comment]
Alright, un-deprecating `posix.ts` and `win32.ts` (and not updating their implementations) in this PR. I'll change `mod.ts` to re-export the single-file exports.

**Comment by lino-levan** [comment]
I guess this module will have quite the turbulent future. This should no longer be a breaking change (even if I personally think it should be). We now have duplicated implementations between `posix.ts` + `win32.ts` and the single-file exports. If we add an OS argument (or export the internal functions), we should be able to consolidate them.

**Comment by lino-levan** [comment]
Updated the original comment for future `os` argument PR. This should be reviewable now.

File: path/basename.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  if (typeof suffix !== "string") {
+    throw new TypeError(
+      `Suffix must be a string. Received ${JSON.stringify(suffix)}`,
+    );
+  }
+
+  if (isWindows) {
+    return windowsBasename(path, suffix);
+  }
+  return posixBasename(path, suffix);
```
Comment (iuioiua) [inline path/basename.ts:61]: Nit: these could be ternaries.

**Review by iuioiua — COMMENTED** [review]
Sidenote: each function in this sub-module should have JSDocs. I guess this can be done in a follow-up PR.

File: path/basename.ts (hunk 2 above)
Comment (lino-levan) [reply path/basename.ts:61]: Yup, fair point.

**Comment by lino-levan** [comment]
With #3515, is there any value in keeping `posix.ts` and `win32.ts`?

**Comment by kt3k** [comment]
It looks like in most test cases, the APIs are called via `path.posix` or `path.win32`. So with this change, new implementations are note tested in CI.

Can you also replace the APIs under path.posix and path.win32 with the new ones?

I think each platform specific function (like `windowsBasename`) should work independently, and we should re-export it as platform specific version.

**Comment by lino-levan** [comment]
I would need to export `windowsX` and `posixX` to do this. Are we committed to doing that?

**Comment by lino-levan** [comment]
Alternatively, which is what I initially pushed for, we could implement an `os` argument, and keep those implementations internal while still being able to test them. Additionally, we would be able to "re-export" the new implementations in `posix.ts` and `win32.ts`.

I don't think we should make our API surface larger over this, especially with the naming as it is.

**Comment by kt3k** [comment]
> I don't think we should make our API surface larger over this, especially with the naming as it is.

I agree.

How about moving platform specific implementations of, for example, `basename` (windowsBasename and posixBasename) to `_basename.ts`? We can use them from `basename.ts`, `posix.ts`, and `win32.ts`, but they are still private from the users

**Comment by lino-levan** [comment]
That seems reasonable. I'll go ahead with that.

**Comment by lino-levan** [comment]
I was skeptical at first, but I'm now convinced that this was the way to do it. Thanks for the suggestion @kt3k! This should now *really* be ready for review.

**Review by iuioiua — APPROVED** [review]
Few nits. Otherwise, LGTM! Nice.

**Review by kt3k — APPROVED** [review]
LGTM. Nice work! @lino-levan 

I'd tag this change as `feat` as newly available paths can be considered as new features.

**Pull request merged**